src/jit/emitxarch.cpp

```diff
@@ -4440,12 +4440,16 @@ void emitter::emitIns_R_R_S_I(
 //
 // Returns:
 //    opReg encoded in imm[7:4]
-static int encodeXmmRegAsIval(regNumber opReg)
+static int8_t encodeXmmRegAsIval(regNumber opReg)
 {
-    assert(opReg >= XMMBASE);
     // AVX/AVX2 supports 4-reg format for vblendvps/vblendvpd/vpblendvb,
     // which encodes the fourth register into imm8[7:4]
-    return (opReg - XMMBASE) << 4;
+
+    assert(opReg >= XMMBASE);
+    int ival = (opReg - XMMBASE) << 4;
+
+    assert((ival >= 0) && (ival <= 255));
```

Comment:
I think the assert would be clearer as `assert((uint8_t)ival == ival);`

Reply:
I'm not convinced that `(uint8_t)ival == ival` is clearer, since it requires you to consider what the range of `uint8_t` is, while the current code explicitly spells it out.

Reply:
@CarolEidt, we actually can't do that and both are required.

The JIT does signed check (value must be in range of `-128 to 127`) for "fits in byte". However, the value returned by the computation, `(opReg - XMMBASE) << 4`, is within the unsigned range (`0 to 255`).

If we cast to `uint8_t`, then the value is zero-extended, rather than sign-extended, and we will fail the "fits in byte" check. This was the same problem that was plaguing the HWIntrinsic APIs that directly take an immediate parameter (as they take a `byte`). -- I think we also discussed, briefly, that this may cause us to emit code that is larger than necessary in some cases (but I don't see an issue tracking that)...